src/crossref.js

```diff
@@ -1,18 +1,272 @@
+        const itemsToBeUpdated = numberOfCitations.filter((number) => number > 0).length;
+        const citationsToBeAdded = numberOfCitations.reduce((sum, value) => sum + value, 0);
+        if (citationsToBeAdded == 0){
+            progress.updateLine(
+                'error',
+                Wikicite.getString('wikicite.crossref.get-citations.no-references')
+            );
+            return;
+        }
+        const confirmed = Services.prompt.confirm(
             window,
-            Wikicite.getString('wikicite.global.unsupported'),
-            Wikicite.getString('wikicite.crossref.get-citations.unsupported')
+            Wikicite.getString('wikicite.crossref.get-citations.confirm-title'),
+            Wikicite.formatString('wikicite.crossref.get-citations.confirm-message', [itemsToBeUpdated, sourceItems.length, citationsToBeAdded])
+        )
+        if (!confirmed){
+            progress.close();
+            return;
+        }
+            
+        // Parse this reference information, then add to sourceItems
+        progress.updateLine(
+            'loading',
+            Wikicite.getString('wikicite.crossref.get-citations.parsing')
+        );
+
+        try {
+            let parsedItemReferences = [];
+            let parsedItems = 0;
+            for (let sourceItemReferenceList of sourceItemReferences) {
+                if (!sourceItemReferenceList.length) {
+                    parsedItemReferences.push([]);
+                    continue;
+                }
+
+                parsedItemReferences.push(await Crossref.parseReferences(sourceItemReferenceList));
+                progress.updateLine(
+                    'loading',
+                    Wikicite.formatString('wikicite.crossref.get-citations.parsing-progress', [++parsedItems, itemsToBeUpdated])
+                );
+            }
+            
+            // Add these citations to the items
+            await Zotero.DB.executeTransaction(async function() {
+                for (let i = 0; i < sourceItemsWithDOI.length; i++){
+                    const sourceItem = sourceItemsWithDOI[i];
+                    const newCitedItems = parsedItemReferences[i];
+                    if (newCitedItems.length > 0){
+                        const newCitations = newCitedItems.map((newItem) => new Citation({item: newItem, ocis: []}, sourceItem));
+                        sourceItem.addCitations(newCitations);
+                    }
+                }
+            })
+            progress.updateLine(
+                'done',
+                Wikicite.getString('wikicite.crossref.get-citations.done')
+            );
+        }
+        catch (error){
+            progress.updateLine(
+                'error',
+                Wikicite.getString('wikicite.crossref.get-citations.error-parsing-references')
+            );
+            debug(error);
+        }
+        finally {
+            progress.close();
+        }
```

Comment:
Yeah I think this would be the nicer user experience for sure.

It should be possible just to drop the items for which `parseReferences` fails, and add the citations to all the rest?

At the moment `addCitations` isn't idempotent, but that's more a problem with citation duplication detection [#25](https://github.com/diegodlh/zotero-cita/issues/25)

Reply:
Eek on #25 — though do duplicates actually cause problems, at this point?

It would be great to store the failures, so we can retry/present to the user, but that should be out of scope for this change.

I vote you (we) go ahead with the change. My own style is to prefer `map`-type code to the explicit loops (with indexing) and I think the functional style of the former makes it easier to review and maintain.

Reply:
Having duplicate citations locally isn't a big deal for the most part. When syncing with Wikidata, creating duplication entries for new items would be a big deal, but we already check and remove duplicates there. But this isn't as issue particular to this PR - you can already add the same citations twice with other methods (by file or identifier).